### PR TITLE
Allow redefinition of where logging goes

### PR DESCRIPTION
**Pull request by thirstyice**
Sometimes I want to log to places other than `Serial` (when I have a display connected, or when my adapter is attached to `Serial1` for example)

Also adds guards so things don't break when the user disables `Serial`

### CODE REVIEW

File: Artnet/Receiver.h [hunk 1]
```diff
@@ -293,12 +293,22 @@ class Receiver_
...
+        log = dest;
+    }
+
 protected:
     void attach(S& s)
     {
         this->stream = &s;
     }
 
+#if defined(NO_GLOBAL_INSTANCES) || defined(NO_GLOBAL_SERIAL)
```
Comment (hideakitai) [inline Artnet/Receiver.h:306]: Could you let me know what these preprocessors are?

File: Artnet/Receiver.h [hunk 2]
```diff
@@ -293,12 +293,22 @@ class Receiver_
...
+    Print* log = nullptr; // No Serial -> no output
```
Comment (hideakitai) [inline Artnet/Receiver.h:307]: I think if the user doesn’t set this, the program crashes. (nullptr->print() will be called)

**Review by hideakitai — COMMENTED** [review]
Thank you, please check the comments.

File: Artnet/Receiver.h (hunk 1 above)
Comment (thirstyice) [reply Artnet/Receiver.h:306]: They're the same macros that allow disabling `Serial`et al in the esp32 arduino core (see [here](https://github.com/arduino/arduino-esp32/blob/21b88659b9ded3fcc1082f23f498bc2a04cd4f1b/cores/esp32/HardwareSerial.cpp#L98C1-L98C64)).
I had assumed when I was developing this that they were standard across all arduino cores, but it seems that they are not.

File: Artnet/Receiver.h (hunk 2 above)
Comment (thirstyice) [reply Artnet/Receiver.h:307]: Upon further research: this is actually undefined behaviour (not neccesarily a crash), which is why I didn't catch it earlier.
Solution coming soon.

**Comment by thirstyice** [comment]
Added a subclass of `Print` that does nothing, which is now the default log output. The user can then choose if + where to log by calling `logOutputTo`.

Removed `b_verbose`, as it seemed redundant to have to call 2 separate functions to enable logging.

This way we also don't have to worry about what happens if `Serial` is undefined, that becomes the user's problem.

**Review by hideakitai — APPROVED** [review]
Almost LGTM. Please check the comment

File: Artnet/Receiver.h [hunk 3]
```diff
@@ -299,6 +284,8 @@ class Receiver_
         this->stream = &s;
     }
 
+    Print* log = &NoLog;
```
[the remark this answers is not in the document]
Comment (hideakitai) [reply Artnet/Receiver.h:287]: Oh well, the default was verbose = false, so it is ok if the default is NoSerial—my mistake.

File: Artnet/Common.h [hunk 4]
```diff
@@ -85,6 +85,13 @@ struct Destination
     uint8_t universe;
 };
 
+class NoPrint : public Print {
+    size_t write(uint8_t) {
+        return 0;
+    }
+};
+NoPrint NoLog;
```
[the remark this answers is not in the document]
Comment (hideakitai) [reply Artnet/Common.h:93]: Also, I think it's better to make `NoLog` to `static NoPrint no_log;` (`static` is for the size optimization for multiple receivers, and renaming to snake case is in terms of variable naming uniformity)

**Pull request merged**